### ISSUE TITLE
Memory corruption? 

### ISSUE THREAD

**Issue by fgsch** [tinygo-org/tinygo#1966]
While testing #1964 with the code further down, I got:

```
021/06/21 23:32:49 net/http: invalid Cookie.Domain "<lots of garbage>"; dropping domain attribute
"foo="
```

The same code compiled with ~`-gc leaking` or~ `-scheduler none` works as expected (just prints `"foo="`).

Code:

```go
package main

import (
        "fmt"
        "net/http"
)

func main() {
        c := http.Cookie{Name:"foo"}
        fmt.Printf("%q\n", c.String())
}
```

Command line:

`tinygo build -target wasi -o test.wasm test.go`

**Comment by fgsch** [comment]
I can reproduced this issue with 0.17.0, 0.18.0 and 0.19.0.

**Comment by fgsch** [comment]
Reduced repro (updated):

```go
package main

import "log"

type foo struct {
	bar string
	qux string
}

func main() {
	func(f *foo) {
		log.Printf("")
		if len(f.bar) != 0 {
			panic("len(f.bar) not 0")
		}
	}(&foo{})
}
```

**Comment by fgsch** [comment]
Fails with 0.17.0 and 0.18.0 as well.

**Comment by aykevl** [comment]
I have further narrowed down this bug to this:

```go
package main

import (
	"os"
	"time"
)

var std = &Logger{out: os.Stderr}

type foo struct {
	a int
	b int
	c int
}

type Writer interface {
	Write([]byte) (int, error)
}

type Logger struct {
	out Writer
}

func main() {
	f(&foo{})
}

func f(f *foo) {
	println("f:", f.a, f.b, f.c)
	time.Time{}.Location().String()
	std.out.Write(nil)
	println("f:", f.a, f.b, f.c)
}
```

It prints the following:

```
$ tinygo run -target=wasi .
f: 0 0 0
f: 0 2 0
```

But clearly it should print the following:

```
$ go run .
f: 0 0 0
f: 0 0 0
```

If you want to help, it would be great if you could break this issue down even further. My strategy so far is inlining functions from packages, seeing whether the copy still reproduces the issue, and then incrementally removing pieces that don't matter while keeping the memory corruption.

**Comment by fgsch** [comment]
I've fiddled with it a bit more and it looks like locking the mutex, and thus interacting with the scheduler, is related. 
Updated code below.

```go
package main

import (
	"os"
	"sync"
)

var std = &Logger{out: os.Stderr}

type foo struct {
	a int
	b int
	c int
}

type Writer interface {
	Write([]byte) (int, error)
}

type Logger struct {
	out Writer
}

func main() {
	f(&foo{})
}

func f(f *foo) {
	var m sync.Mutex
	println("f:", f.a, f.b, f.c)
	m.Lock()
	std.out.Write(nil)
	println("f:", f.a, f.b, f.c)
}
```

 To be continued.

**Comment by fgsch** [comment]
This might be the same as in #1981.

**Comment by aykevl** [comment]
Maybe, although my first guess is that they're unrelated. WebAssembly does some things very differently from other instruction sets (such as amd64) and it requires some ugly workarounds in TinyGo that I suspect cause this bug.

**Comment by aykevl** [comment]
You were right after all, https://github.com/tinygo-org/tinygo/pull/1999 also fixes this issue.

**Comment by deadprogram** [comment]
Released with 0.20.0 so now closing. Thank you!
